modules/ravestate_sendpics/__init__.py

```diff
@@ -0,0 +1,98 @@
+from ravestate.module import Module
+from ravestate.state import s, state, Emit
+from ravestate.property import PropertyBase
+from ravestate.wrappers import ContextWrapper
+import ravestate_ontology
+import face_recognition as fr
+import pickle
+import redis
+
+import ravestate_rawio
+import ravestate_nlp
+import ravestate_idle
+
+from scientio.ontology.node import Node
+from scientio.session import Session
+from scientio.ontology.ontology import Ontology
+
+from reggol import get_logger
+logger = get_logger(__name__)
+
+REDIS_HOST_CONF = "redis_host"
+REDIS_PORT_CONF = "redis_port"
+REDIS_PASS_CONF = "redis_pass"
+CONFIG = {
+    REDIS_HOST_CONF: "localhost",
+    REDIS_PORT_CONF: 6379,
+    REDIS_PASS_CONF: None
+}
+
+
+with Module(name="sendpics", config=CONFIG):
+
+    face_vec = PropertyBase(name="face_vec", always_signal_changed=True)
+
+    @state(cond=s("idle:bored"), write="rawio:out")
+    def prompt_send(ctx):
+        ctx["rawio:out"] = "Why don't you send me a picture? I'm really good at recognizing faces!"
+
+    @state(
+        read="rawio:pic_in",
+        write=("rawio:out", "sendpics:face_vec"),
+        emit_detached=True)
+    def prompt_name(ctx):
+        # Get face ancoding
+        image = fr.load_image_file(ctx["rawio:pic_in"])
+        faces = fr.face_encodings(image)
+        # Prompt name
+        if len(faces) > 0:
+            if len(faces) > 1:
+                ctx["rawio:out"] = f"Nice! I am seeing {len(faces)} people in that picture. Who is the first one?"
+            else:
+                ctx["rawio:out"] = "Nice picture! Who is that person?"
+            ctx["sendpics:face_vec"] = pickle.dumps(faces[0])
+        else:
+            ctx["rawio:out"] = f"Huh, looks interesting. But maybe send me a cute face!"
+
+    @state(
+        cond=(
+            s("sendpics:repeat_name", max_age=30.) | s("sendpics:face_vec:changed", max_age=30.)
+        ) & s("nlp:tokens:changed"),
+        signal_name="repeat_name",
+        read=("nlp:triples", "nlp:tokens", "sendpics:face_vec"),
+        write="rawio:out",
+        emit_detached=True)
+    def store_face_and_name(ctx: ContextWrapper):
+        tokens = ctx["nlp:tokens"]
+        triples = ctx["nlp:triples"]
+        if len(tokens) == 1:
+            name = tokens[0]
+        elif triples[0].get_object().text and triples[0].match_either_lemma(pred={"be"}):
```

Comment:
is there a case when len(triples) is 0 after nlp:tokens:changed?

Reply:
There is always at leats one guaranteed triple